Allow users to have multiple roles and removes banned role

DubeySandeep opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->
1. This PR does the following: Allows users to have multiple roles and have a separate field to mark a banned user.
More info in [design doc](https://docs.google.com/document/u/1/d/1xp-HvyVut4ean3p-aAKGQZng0czMYw1K9PHasNg8eXk/edit#heading=h.liiqbmwmz8j8)

Some other minor changs in this PR: 
- Changing actions assigned to some roles.
- Replacing Admin role with Curriculum admin

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes made in this PR work correctly.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->

New UI for assigning multiple roles to user: https://drive.google.com/file/d/1ymGmvyux2YkxfyzkBmIJzJ_sAEE3uQXG/view

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

DubeySandeep committed:
Adds one-off job for populating roles field

DubeySandeep committed:
Fix backend test issue

DubeySandeep committed:
Updated with upstream/develop

DubeySandeep committed:
Updates job with new design

DubeySandeep committed:
Merge branch 'develop' of github.com:oppia/oppia into roles-job

DubeySandeep committed:
Updates functionality

DubeySandeep committed:
Adds TODO

DubeySandeep committed:
Fix dosctring

DubeySandeep committed:
Cleanup

DubeySandeep committed:
Adds more TODO comment

DubeySandeep committed:
Updates TODO comment

DubeySandeep committed:
Adds export policy for new fields

DubeySandeep committed:
Remove unwanted checks

DubeySandeep committed:
Address review comments

DubeySandeep committed:
Adds choices for roles property

DubeySandeep committed:
Fix test issue

DubeySandeep committed:
Updates TODO comments

DubeySandeep committed:
Merge branch 'develop' of github.com:oppia/oppia into roles-job

DubeySandeep committed:
Update with upstream/develop

DubeySandeep committed:
Merge branch 'develop' of github.com:oppia/oppia into roles-job

DubeySandeep committed:
Use roles field for user roles.

DubeySandeep committed:
Resolves merge conflicts

DubeySandeep committed:
Fix conflict issues

DubeySandeep committed:
Removes shard to get all beckend test issues

DubeySandeep committed:
Resolve ts lint issues

DubeySandeep committed:
Fix lint issues

DubeySandeep committed:
Remove shard

DubeySandeep committed:
Fix lint issues

DubeySandeep committed:
Fix e2e test issue

DubeySandeep committed:
Removes fix param

DubeySandeep committed:
Fix test issues

DubeySandeep committed:
replase admin_username with curriculum admin username

DubeySandeep committed:
Replace admin email with curriculum admin email

DubeySandeep committed:
Fix name issue

DubeySandeep committed:
Fix tests

DubeySandeep committed:
Fix conflicts

DubeySandeep committed:
Fix conflicts

DubeySandeep committed:
Fix tests

DubeySandeep committed:
Fix tests

DubeySandeep committed:
Fix test and rename

DubeySandeep committed:
Merge branch 'develop' of github.com:oppia/oppia into use-roles

DubeySandeep committed:
Fix lint issues

DubeySandeep committed:
Fix test issues

DubeySandeep commented on core/controllers/base.py:
```diff
@@ -227,14 +227,14 @@ def __init__(self, request, response):  # pylint: disable=super-init-not-called
-        self.values['is_moderator'] = (
-            user_services.is_at_least_moderator(self.user_id))
-        self.values['is_admin'] = user_services.is_admin(self.user_id)
+        self.values['is_moderator'] = user_services.is_moderator(self.user_id)
+        self.values['is_curriculum_admin'] = (
+            user_services.is_curriculum_admin(self.user_id))
```
@ashutoshc8101  Planning to remove all this seperate is_* key values and only send user_roles through the API and use it as needed in the frontend.

DubeySandeep committed:
Self review address

DubeySandeep committed:
Resolve merge conflicts

DubeySandeep committed:
Fix minor issue

DubeySandeep committed:
Fix test issue

DubeySandeep committed:
Fix test issue

DubeySandeep committed:
Fix test issue

DubeySandeep committed:
Resolve conflicts

DubeySandeep committed:
Fix backend test issue

DubeySandeep committed:
Fix test issue

DubeySandeep committed:
Fix tests

DubeySandeep committed:
Removes shard for full testing

DubeySandeep committed:
Fix tests

DubeySandeep committed:
Fix conflict

DubeySandeep committed:
Fix test issues

DubeySandeep committed:
Fix tests

DubeySandeep committed:
Resolve conflicts

DubeySandeep committed:
fix test

DubeySandeep committed:
Fix issues

DubeySandeep committed:
Resolve merge conflicts

DubeySandeep committed:
Address review comments

DubeySandeep committed:
Fix lint issue

seanlip commented on core/controllers/admin.py:
```diff
@@ -205,10 +205,7 @@ def get(self):
-                role: role_services.HUMAN_READABLE_ROLES[role]
-                for role in role_services.VIEWABLE_ROLES
-            },
+            'viewable_roles': role_services.VIEWABLE_ROLES,
```
Shouldn't this and updatable_roles be in the same format for parity reasons?

DubeySandeep replied:
I have send it the way I want to use it in the frontedend, I've undo the changes to keep the API as-is.

seanlip commented on core/controllers/admin.py:
```diff
@@ -741,37 +763,125 @@ def get(self):
+            topic_ids = []
+            if feconf.ROLE_ID_TOPIC_MANAGER in user_roles:
+                topic_ids = [
+                    t_r.id for t_r in topic_fetchers.get_topic_rights_with_user(
```
I think it's OK to use `topic_rights` (or perhaps `rights`) here instead of `t_r`. Just break after 'in'.

DubeySandeep replied:
Done! (I had seen this pattern multiple places in list compression, so just followed it but I think making it topic_right is fine!)

seanlip commented on core/controllers/admin.py:
```diff
@@ -712,6 +728,12 @@ class AdminRoleHandler(base.BaseHandler):
+            },
+            'remove_from_all_topics': {
+                'schema': {
+                    'type': 'basestring'
```
If I'm understanding this correctly, shouldn't it be a boolean (conceptually)? /cc @Nik-09

DubeySandeep replied:
We cannot send boolean through query_string in the forntend. Typescript doesn't allow me to do so, I'm doing this booleanVar.toString()

DubeySandeep replied:
I've added choices to be `'true'` or `'false'`*

seanlip replied:
But shouldn't the server interpret it as a boolean? /cc @Nik-09

DubeySandeep replied:
Removed as the field is not needed!

seanlip commented on core/controllers/admin.py:
```diff
@@ -741,37 +763,125 @@ def get(self):
+
+            user_settings = user_services.get_user_settings(user_id)
+            user_roles = user_settings.roles
+            topic_ids = []
```
Maybe make more explicit what this is, e.g. managed_topic_ids. (Ditto for the key in the returned dict.)

DubeySandeep replied:
Done!

seanlip commented on core/controllers/admin.py:
```diff
@@ -741,37 +763,125 @@ def get(self):
+    def put(self):
+        username = self.payload.get('username')
+        role = self.payload.get('role')
+        topic_id = self.payload.get('topic_id')
```
Again, give this a more descriptive name that better indicates what it conceptually represents and why it makes sense for it to be passed to an "admin roles handler".

DubeySandeep replied:
I've followed the existing pattern i.e, we have topic_id in the current API. I was trying not to alter the API from clients perspective (Ditto for other places where I've used topic_id(s)). I didn't thought of renaming it even I found it to be not a good name, I can change it if needed. If yes, do you have any sugestion?

seanlip replied:
I think the problem is that this handler does a bunch of things that may not be related to topics. So you'll need to explain the connection in the name.

If you're having trouble coming up with a name, explain clearly what this variable represents in long-form and I can help you condense it.

DubeySandeep replied:
Created seperate handler for assigning/deassign user from topic.

seanlip commented on core/controllers/admin.py:
```diff
@@ -741,37 +763,125 @@ def get(self):
+    def delete(self):
+        username = self.request.get('username')
+        role = self.request.get('role')
+        topic_id = self.request.get('topic_id')
```
Same thing.

DubeySandeep replied:
This is an outdated code, create seperate handler to handle topic assignment.

seanlip commented on core/controllers/admin.py:
```diff
@@ -741,37 +763,125 @@ def get(self):
+        role = self.request.get('role')
+        topic_id = self.request.get('topic_id')
+        remove_from_all_topics = self.request.get(
+            'remove_from_all_topics', False)
```
Noting that the default value here is a boolean, not a string ... but you shouldn't be doing that, anyway, right? The default value should be specified in the schema framework. /cc @Nik-09

DubeySandeep replied:
> but you shouldn't be doing that, anyway, right? The default value should be specified in the schema framework.

I had these changes before the schema_validation structure was implement and this got left as-is. Fixed!

seanlip commented on core/controllers/admin.py:
```diff
@@ -741,37 +763,125 @@ def get(self):
-            self.user_id, feconf.ROLE_ACTION_UPDATE, role=role,
-            username=username)
+        topic_ids = [
+            t_r.id for t_r in topic_fetchers.get_topic_rights_with_user(
```
Same note as above re t_r

DubeySandeep replied:
Done!

seanlip commented on core/controllers/admin.py:
```diff
@@ -741,37 +763,125 @@ def get(self):
+                user_id)]
+        if topic_id not in topic_ids:
+            raise self.InvalidInputException(
+                'User does no have rights in the give topic.')
```
There are two typos in this error message.

DubeySandeep replied:
Fixed!

seanlip commented on core/controllers/admin.py:
```diff
@@ -741,37 +763,125 @@ def get(self):
+            self.render_json({})
+            return
+
+        if remove_from_all_topics:
```
Why do we need to pass a separate bool and rely on the client for this? I think we should check the topic list after doing any operations and remove the user role then if needed.

I also feel like there are too many "independent" if-conditions here and that we should have more of an if/else structure to make sure all cases are covered in a mutually exclusive way. The exception is for "special case" stuff at the top of the method but "topic manager" isn't a special case.

DubeySandeep replied:
> Why do we need to pass a separate bool and rely on the client for this? I think we should check the topic list after doing any operations and remove the user role then if needed.

We are doing the check internally and we are not relying on client. But in case client wants to remove the TOPIC_MANAGER role completely from the user we are sending remove_from_all_topics=True.

Current strucutre:
```
To remove from one topic (can remove topic_manager role if user only manages 1 topic):
role: TOPIC_MANGER
topic_id: <topicId>

To remove from all topics and remove topic_manager role:
role: TOPIC_MANGER
topic_id: <topicId>
```
Another possible solution:
```
To remove from one topic (can remove topic_manager role if user only manages 1 topic):
remove_from_topic_id: <topicId>

To remove from all topics and remove topic_manager role:
role: TOPIC_MANGER
```

Let me know whether the "another" solution sounds okay? (I'm trying to make it either remove 1 or all, as we are handling one role editing at a time through this handler.)

seanlip replied:
I would say that the cleanest option is: for any operations related to the topic manager role, they must **always** include a topic ID. The end user never manages that role directly.

You might even go so far as to have an entirely separate handler for operations related to topic management.

DubeySandeep replied:
Create seperate handler for topic assignment/deassignment.

seanlip commented on core/controllers/admin.py:
```diff
@@ -741,37 +763,125 @@ def get(self):
+        topic_services.deassign_manager_role_from_topic(
+            user_services.get_system_user(), user_id, topic_id)
+
+        if len(topic_ids) == 1:
```
Why are we doing this?

DubeySandeep replied:
In case before removal user has rights only in one topic then remove the TOPIC_MANGER role from user's roles.

DubeySandeep replied:
Should I add a codecomment? or do a `topic_ids.remove(topic_id_to_remove)` and then check `if not topic_ids`?

seanlip replied:
See comment above for how to simplify it. (But here you'd probably want to be referring to the final list of topic IDs, not the initial one.)

DubeySandeep replied:
Done! (Created seperate handler)

seanlip commented on core/controllers/admin.py:
```diff
@@ -741,37 +763,125 @@ def get(self):
+
+
+class BannedUsersHandler(base.BaseHandler):
+    """Handler mark/unmark a banned user."""
```
Please write this docstring grammatically.

DubeySandeep replied:
Done!

seanlip commented on core/controllers/admin_test.py:
```diff
@@ -1163,8 +1199,120 @@ def test_changing_user_role_from_exploration_editor_to_topic_manager(self):
+            })
+        self.logout()
+
+    def test_adding_new_topic_to_a_topic_manger(self):
```
'manager' is misspelled.

DubeySandeep replied:
Done!

seanlip commented on core/controllers/moderator_test.py:
```diff
@@ -43,13 +43,6 @@ def test_moderator_page(self):
         self.get_html_response('/moderator')
         self.logout()
 
-        # Try accessing the moderator page after logging in as an admin.
```
Why not still test this (and assert 404)?

DubeySandeep replied:
Earlier we had this testbecause we were expecting admin to access the page.

Asserting 404 will be same as assert other users. I don't wanted to assert just for curriculum-admin else it won't make sense that why we haven't asserted it for translation-admin, voiceover-admin, question admin <other roles> etc.

A better way to test these handlers woul be to mock the acl , but we don't have that structure! (And unit test for the acl should mock the actions for a role and ensure taht the user with following allowed actions can pass the check!)

seanlip commented on core/controllers/payload_validator.py:
```diff
@@ -83,6 +83,7 @@ def validate(handler_args, handler_args_schemas, allowed_extra_args):
     'AnswerSubmittedEventHandler',
     'AssetDevHandler',
     'AudioUploadHandler',
+    'BannedUsersHandler',
```
Shouldn't this be removed and you fix this in this PR, since it's new? /cc @Nik-09

DubeySandeep replied:
I've already have the schmea, I think I missed removing it from this file* (I had added it when the payload validator was not working fine on develop)

DubeySandeep replied:
Removed!

seanlip commented on core/controllers/topic_editor_test.py:
```diff
@@ -312,10 +312,6 @@ def test_editable_subtopic_page_get(self):
         }, json_response['subtopic_page']['page_contents'])
         self.logout()
 
-        topic_services.assign_role(
```
Could you explain why this, or an equivalent statement, is no longer needed?

DubeySandeep replied:
This is a common setup procces in all the tests so I've mode it to `BaseTopicEditorControllerTests.setUp` as we have done all the role assignment in the setUP function* (line 84 in new changes)

seanlip commented on core/controllers/topic_editor_test.py:
```diff
@@ -791,10 +787,6 @@ def test_editable_topic_handler_put_for_assigned_topic_manager(self):
                 'subtopic_id': 2
             }]
         }
-        # Assign the topic manager to the topic.
```
Ditto.

DubeySandeep replied:
Explained above*

seanlip commented on core/controllers/topics_and_skills_dashboard_test.py:
```diff
@@ -133,7 +134,7 @@ def test_get(self):
         self.assertEqual(len(json_response['topic_summary_dicts']), 1)
         self.assertEqual(
             json_response['topic_summary_dicts'][0]['can_edit_topic'],
-            False)
```
Can you explain why this changed?

DubeySandeep replied:
It's a mistake, I had addressed this earlier but had not pushed the self-review comments before your review comments. Fixed!

seanlip replied:
No, it's not fixed...

DubeySandeep replied:
I looked into it again, I found this to be a correct change i.e, earlier our API was not working fine and we were allowing assigning topic_manger without assigning them to any topic (see line 50 in old file). Now I ahve assigne the user to the topic with the topic id so they should have access to edit the topic! (Also, written on line 129 i.e, the user should have acces to editable topic data.)

seanlip commented on core/domain/config_domain.py:
```diff
@@ -172,7 +172,7 @@ class ConfigProperty(python_utils.OBJECT):
     - admin_email_address.
     - admin_ids.
-    - admin_usernames.
+    - curriculum_admin_usernames.
```
Why is this changed?

When looking at this, I am concerned that you might have just done a find/replace indiscriminately throughout the codebase. Please check all changes carefully and verify that they are correct.

DubeySandeep replied:
> When looking at this, I am concerned that you might have just done a find/replace indiscriminately throughout the codebase. Please check all changes carefully and verify that they are correct.

Yep, that's somewhat true i.e, I did a `find `and have gone through each places and allowed replacement there if needed. I assumed that it should be fine to rename it here so taht we have similar name in all the places. But I think this should be kept as-is to avoid introducing new config with similar name*

Removed!

seanlip commented on core/domain/email_manager.py:
```diff
@@ -337,9 +336,8 @@ def log_new_error(*args, **kwargs):
-    feconf.EMAIL_INTENT_MARKETING: user_services.is_admin,
-    feconf.EMAIL_INTENT_DELETE_EXPLORATION: (
-        user_services.is_at_least_moderator),
+    feconf.EMAIL_INTENT_MARKETING: user_services.is_curriculum_admin,
```
I don't see why a curriculum admin should be sending marketing emails. Is there another role (e.g. super admin or something) that would be more appropriate for this?

DubeySandeep replied:
I've changed it to `feconf.SYSTEM_COMMITTER_ID`

seanlip commented on core/domain/email_manager.py:
```diff
@@ -358,14 +356,18 @@ def log_new_error(*args, **kwargs):
-    feconf.BULK_EMAIL_INTENT_IMPROVE_EXPLORATION: user_services.is_admin,
-    feconf.BULK_EMAIL_INTENT_CREATE_EXPLORATION: user_services.is_admin,
-    feconf.BULK_EMAIL_INTENT_CREATOR_REENGAGEMENT: user_services.is_admin,
+    feconf.BULK_EMAIL_INTENT_MARKETING: user_services.is_curriculum_admin,
```
Ditto for this one (marketing), as well as: creator reengagement, learner reengagement, and "test".

DubeySandeep replied:
Done!

seanlip commented on core/controllers/base.py:
```diff
@@ -229,18 +229,18 @@ def __init__(self, request, response):  # pylint: disable=super-init-not-called
-            user_services.is_at_least_moderator(self.user_id))
-        self.values['is_admin'] = user_services.is_admin(self.user_id)
+        self.values['is_moderator'] = user_services.is_moderator(self.user_id)
+        self.values['is_curriculum_admin'] = (
```
Do we actually need this?

DubeySandeep replied:
Not sure but I've tried to keep the API structure as-is!

seanlip replied:
I would suggest checking to see where it's used. This may uncover cases where we're looking for an "admin"-type person but which no longer make sense with a "curriculum admin". Ideally we get rid of it completely.

DubeySandeep replied:
Looks like it's not been used anywhere in eth UI but we pass it in reponse for most of the handlers. I can do this cleanup if needed in a seperate PR! (Let me know if it's okay to do it ina  seperate PR then I'll create an issue!)
Few examples:
![image](https://user-images.githubusercontent.com/16653571/126416491-a3e0e9a8-5ee5-4618-82e9-bd61109a7c79.png)
![image](https://user-images.githubusercontent.com/16653571/126416508-1b29ce89-cc26-4d44-b3d4-16896a3836bb.png)

Note: We also send is_super_admin and is_moderator*

seanlip replied:
I would prefer to drop it here, but if you would prefer a separate PR then please file an issue and add a TODO.

seanlip commented on core/domain/role_services.py:
```diff
@@ -96,226 +96,188 @@
+        ACTION_EDIT_SKILL,
+        ACTION_EDIT_SKILL_DESCRIPTION,
+        ACTION_EDIT_OWNED_TOPIC,
+        ACTION_MANAGE_EMAIL_DASHBOARD,
```
Perhaps this one should be super admin? Not sure how it relates to curriculum admin.

DubeySandeep replied:
We don't have roles_action for super_admin as super admin is not a user role the app handles, it's more of a role being handleed by firebase internally*

For now I've changed the `can_manage_email_dashboard` acl to only allow super-admin, is that fine?

seanlip replied:
Yup

DubeySandeep replied:
Done!

seanlip commented on core/domain/role_services.py:
```diff
@@ -96,226 +96,188 @@
+            raise Exception('Role %s does not exist.' % role)
 
-    role_actions = _ROLE_ACTIONS[role]
+        role_actions |= set(_ROLE_ACTIONS[role])
```
Why not just do role_Actions.add(...)?

DubeySandeep replied:
I think add doesn't accepts list/set i.e, _ROLE_ACTIONS[role] returns a `list of action` and we need to do a union here.

seanlip commented on core/domain/user_domain.py:
```diff
@@ -86,7 +85,8 @@ def __init__(
             email: str. The user email.
-            role: str. Role of the user.
+            roles: list(str). Roles of the user.
+            banned: bool. Whetehr the uses is banned.
```
Whether is misspelled.

DubeySandeep replied:
Done!

seanlip commented on core/domain/user_domain.py:
```diff
@@ -177,11 +178,41 @@ def validate(self):
+                if role in feconf.POSSIBLE_REGISTERED_USER_DEFAULT_ROLES:
+                    defalut_roles.append(role)
+
+            if len(defalut_roles) != 1:
```
Wrong spelling of default_roles.

DubeySandeep replied:
Fixed!

seanlip commented on feconf.py:
```diff
@@ -1045,30 +1045,33 @@ def get_empty_ratings():
 # Following are the constants for the role IDs.
-ROLE_ID_ADMIN = 'ADMIN'
-ROLE_ID_BANNED_USER = 'BANNED_USER'
+# TODO(#<id>): The role id variable name doesn't match the string value,
```
Missing numeric ID.

DubeySandeep replied:
Done!

seanlip commented on feconf.py:
```diff
@@ -1045,30 +1045,33 @@ def get_empty_ratings():
 ROLE_ID_TRANSLATION_ADMIN = 'TRANSLATION_ADMIN'
 ROLE_ID_VOICEOVER_ADMIN = 'VOICEOVER_ADMIN'
 
+POSSIBLE_REGISTERED_USER_DEFAULT_ROLES = [
```
I don't entirely understand the naming here -- what is the difference between a "registered user" and a "user"?

Would ALLOWED_DEFAULT_USER_ROLES_ON_REGISTRATION or similar be sufficient/clearer?

DubeySandeep replied:
Guest users are not a registered user and they don't have default roles. 

Changed it to `ALLOWED_DEFAULT_USER_ROLES_ON_REGISTRATION`

seanlip commented on core/domain/user_domain.py:
```diff
@@ -177,11 +178,41 @@ def validate(self):
+                if not isinstance(role, python_utils.BASESTRING):
+                    raise utils.ValidationError(
+                        'Expected roles to be a string, received %s' %
+                        self.pin
```
This interpolated variable seems wrong.

DubeySandeep replied:
Fixed!

seanlip commented on core/domain/user_domain.py:
```diff
@@ -386,19 +418,33 @@ def require_valid_username(cls, username):
+        Args:
+            default_role: str. The role assigned to the user after marking
+                unbanned.
+        """
```
Should you verify that default_role is actually a default role?

DubeySandeep replied:
It is done in the validation method. (General we don't do such internal validation in the update/action method of the domain object we leave it in the validation method, I see this pattern in all other places.)

seanlip commented on core/domain/user_services.py:
```diff
@@ -1236,14 +1238,42 @@ def update_user_role(user_id, role):
     user_settings = get_user_settings(user_id, strict=True)
-    if user_settings.role == feconf.ROLE_ID_LEARNER:
+    if feconf.ROLE_ID_MOBILE_LEARNER in user_settings.roles:
         raise Exception('The role of a Learner cannot be changed.')
```
Should this message say "Android mobile learner" or similar?

DubeySandeep replied:
I'll change it to mobile learner.

DubeySandeep replied:
Done!

seanlip commented on core/domain/user_services.py:
```diff
@@ -1236,14 +1238,42 @@ def update_user_role(user_id, role):
+    """
+    user_settings = get_user_settings(user_id, strict=True)
+    if feconf.ROLE_ID_MOBILE_LEARNER in user_settings.roles:
+        raise Exception('The role of a Learner cannot be changed.')
```
Ditto.

DubeySandeep replied:
Done!

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -192,10 +193,9 @@ def get_export_policy(cls):
-            # "EXPORTED" once the fields are populated in the datastore.
-            'roles': base_models.EXPORT_POLICY.NOT_APPLICABLE,
-            'banned': base_models.EXPORT_POLICY.NOT_APPLICABLE
+            # The role is a deprecated field and doesn't contains any correct
```
contains --> contain

DubeySandeep replied:
Fixed!

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -2621,9 +2622,7 @@ class PendingDeletionRequestModel(base_models.BaseModel):
         datastore_services.StringProperty(indexed=True))
-    # Role of the user. Needed to decide which storage models have to be deleted
-    # for it.
-    role = datastore_services.StringProperty(required=True, indexed=True)
```
I'm not sure you can just delete this, you have to deprecate it first and then delete it later I believe (because there are still models in the datastore with it). See what we did with skill_tags last time (I think @vojtechjelinek and @rohitkatlaa know the process).

DubeySandeep replied:
I've had a conversation with Vojetch and he confirmed that we don't ahve any model in the datastore and this field can be removed! (In general we rarely use to have any such model in datastore as we delete the model once the account is deleted)

@vojtechjelinek Can you confirm?

seanlip replied:
We'd probably need to add an explicit step/check in the release containing this fix **prior to** deployment, right? (To double-confirm that no such model exists.)

DubeySandeep replied:
We delete this model once the deletion request is completed and we also have a sanity check for the same. In any case if there  remains one-two models which are not deleted but expected to be delete won't create any issue.

@vojtechjelinek Can you confirm whether it's fine to remove this field directly?

seanlip commented on core/templates/pages/admin-page/admin-page.component.html:
```diff
@@ -23,8 +23,7 @@
   </admin-features-tab>
 
-  <oppia-admin-roles-tab *ngIf="isRolesTabOpen()"
-                         (setStatusMessage)="setStatusMessage($event)">
```
Why has this changed?

DubeySandeep replied:
The satus message is not required in roles-tab*

seanlip commented on core/templates/pages/admin-page/roles-tab/admin-roles-tab.component.html:
```diff
@@ -3,130 +3,176 @@
+          <mat-icon aria-label="Example icon-button with a heart icon">edit</mat-icon>
+        </button>
+      </div>
+      <div *ngIf="!username">Please fillout the username above to editor the user role.</div>
```
Please fill in the username above in order to edit the user role.

DubeySandeep replied:
I had fixed this in the self-review sesion, had not pushed the changes before your review!

DubeySandeep replied:
Fixed*

seanlip commented on core/templates/pages/admin-page/roles-tab/admin-roles-tab.component.html:
```diff
@@ -3,130 +3,176 @@
     </div>
   </div>
-</div>
+</div>
```
Missing newline at EOF

DubeySandeep replied:
Fixed!

seanlip commented on core/templates/pages/admin-page/roles-tab/admin-roles-tab.component.ts:
```diff
@@ -13,159 +13,173 @@
+ angular.module('oppia').directive('oppiaAdminRolesTab',
+   downgradeComponent({
+     component: AdminRolesTabComponent
+   }) as angular.IDirectiveFactory);
```
Ditto

DubeySandeep replied:
Fixed!

seanlip reviewed (changes requested):
@DubeySandeep This PR is way too big, and is very hard to review correctly. It will likely take at least a couple of weeks at this rate. You might want to consider doing the renaming in one PR (or separate PRs for each renaming) and the multiple roles in another.

For UI -- overall it's quite nice! But a few notes:
- Could you arrange the roles alphabetically in the dropdown?
- Why is it called "exploration editor" in the dropdown and "full user" in the codebase?
- It would be nice to have some sort of confirmation (like a toast at the bottom right) that an action has actually happened. Currently it's hard to tell -- the UI is updated but it's not clear that changes are "saved".

DubeySandeep committed:
Init

DubeySandeep committed:
Address review comments

DubeySandeep commented:
> This PR is way too big

You mean in terms of "number of files" changed? [I've created mutiple such refactor PRs around ~2000 lines changes]

> You might want to consider doing the renaming in one PR (or separate PRs for each renaming) and the multiple roles in another.

The "admin" name changes is the major one other than that other action changes are just in roles_services. Also in most palces where I've changed admin to a curriculum_admin I've also changed the beaviour as now curicuulum admin have different rights. So I don't think having separate PR for name changing would make much different in review experience (Also, would be very tedious work for me to safely undo the changes + safely redo the changes in another PR!).  

For UI -- overall it's quite nice! But a few notes:

> Could you arrange the roles alphabetically in the dropdown?
I had kept it as-is but I think it would be better to show then in sorted order. FIxed!
 
> Why is it called "exploration editor" in the dropdown and "full user" in the codebase?

I've fixed it! (SOme how the human redable test and actual role string values were same!)

> It would be nice to have some sort of confirmation (like a toast at the bottom right) that an action has actually happened. Currently it's hard to tell -- the UI is updated but it's not clear that changes are "saved".

Locally it's hard to see the loading icon as the server is fase but we are showing a loading icon in place of "X" until the roles is added/removed!

Here's the UI for slow connection: https://drive.google.com/file/d/1RDCAVpVYPuYZHAGoWXvTk--CnRMItsGx/view

I can add the the toast if needed but as we are making http call on each actions and we don't have a explicit "save" button, I thought having a toast would look messy when adding/removing multiple roles to/from user.

DubeySandeep commented on core/controllers/acl_decorators.py:
```diff
@@ -453,7 +453,7 @@ def test_can_manage_emails(self, **kwargs):
             raise base.UserFacingExceptions.NotLoggedInException
 
-        if role_services.ACTION_MANAGE_EMAIL_DASHBOARD in self.user.actions:
+        if self.current_user_is_super_admin:
```
Removed this action as per discussion [here](https://github.com/oppia/oppia/pull/12989#discussion_r667256264).

Super-admin role doesn't have role_to_actions structure and we only want super amdins to access the email dashboard page!

seanlip commented:
> You mean in terms of "number of files" changed? [I've created mutiple such refactor PRs around ~2000 lines changes]

Mainly in terms of this PR containing multiple different changes that are not exactly related. The interleaving of different changes makes it harder to reason about the "making roles multi" system, which would be better to review in isolation.

For this PR, it's up to you, though. Just noting that the current way might actually take longer. 

But, that said: please do bear the above in mind for future PRs. It really does make things harder for the reviewers when you have a lot of complex changes mixed together, and it is reasonable for reviewers to request/require that the PR be split into multiple ones in such cases.

> I can add the the toast if needed but as we are making http call on each actions and we don't have a explicit "save" button, I thought having a toast would look messy when adding/removing multiple roles to/from user.

I'm fine with either a toast, or alternatively you can replace the loading icon with a green checkmark that shows up for 5 seconds afterwards, and then disappears. Otherwise, if the loading icon disappears immediately, it's hard to see. Wdyt?

DubeySandeep committed:
Address self review comments

DubeySandeep committed:
Fix test

DubeySandeep committed:
Run all test at once

DubeySandeep committed:
remove shard

DubeySandeep committed:
Refactor e2e tests

DubeySandeep committed:
Use add_role

DubeySandeep committed:
Merge branch 'develop' of github.com:oppia/oppia into use-roles

DubeySandeep committed:
Fix backend tests and lint checks

DubeySandeep committed:
Resolve conflicts

DubeySandeep committed:
Fix test issue

DubeySandeep committed:
Fix test issue

DubeySandeep committed:
Fix e2e test issues

DubeySandeep committed:
Adds shards

DubeySandeep committed:
Adds new topicManager handler

DubeySandeep committed:
Fix test issue

DubeySandeep commented:
> I'm fine with either a toast, or alternatively you can replace the loading icon with a green checkmark that shows up for 5 seconds afterwards, and then disappears. Otherwise, if the loading icon disappears immediately, it's hard to see. Wdyt?

![image](https://user-images.githubusercontent.com/16653571/126418433-20b4d756-b7b6-4c2e-9795-527f903dd2d0.png)

Are you suggesting something around this^ for 5 secs? If yes, not sure what to do for role removal case. 

Can we show a "changes save" mesage at teh bottom of the card for 5 secs?
![image](https://user-images.githubusercontent.com/16653571/126419942-5c38e2b9-205f-4ba2-a5b0-b89703db7647.png)

seanlip commented on core/controllers/admin.py:
```diff
@@ -766,14 +754,14 @@ def get(self):
-                    t_r.id for t_r in topic_fetchers.get_topic_rights_with_user(
-                        user_id)]
+                managed_topic_ids = [
+                    topic_right.id for topic_right in
```
Use "rights" instead of "topic_right". The singular "right" doesn't really make sense here.

DubeySandeep replied:
Done!

seanlip commented on core/controllers/admin.py:
```diff
@@ -782,88 +770,122 @@ def get(self):
-            if feconf.ROLE_ID_TOPIC_MANAGER not in user_settings.roles:
-                user_services.add_user_role(user_settings.user_id, role)
+            raise self.InvalidInputException(
+                'Unsupported role for the handler.')
```
the --> this

Also, maybe append "please assign a specific topic ID instead" or similar. (Alternatively, leaving a comment in the code to explain is also fine.)

DubeySandeep replied:
changed `the` to `this` and added a code-comment

seanlip commented on core/controllers/admin.py:
```diff
@@ -782,88 +770,122 @@ def get(self):
+            topic_services.assign_role(
+                user_services.get_system_user(),
+                topic_manager, topic_domain.ROLE_MANAGER, topic_id)
+        else:
```
elif action == 'deassign'

DubeySandeep replied:
Done!

seanlip commented on core/controllers/admin.py:
```diff
@@ -875,12 +897,13 @@ def put(self):
         if user_id is None:
             raise self.InvalidInputException(
                 'User with given username does not exist.')
+        topic_services.deassign_user_from_all_topics(self.user, user_id)
```
Why is this here? I don't understand the logic behind it.

DubeySandeep replied:
We need unassign user from all the topics while marking ban so that once we mark them unbanned they don't have topic_manger role.

seanlip replied:
But isn't this the wrong handler? It seems to be the one for _unbanning_ a user.

DubeySandeep replied:
Sorry for the confusion, I've fixed it now!

seanlip commented on core/controllers/admin_test.py:
```diff
@@ -1203,11 +1205,11 @@ def test_adding_topic_manager_role_to_user(self):
         self.logout()
 
-    def test_adding_new_topic_to_a_topic_manger(self):
+    def test_adding_new_topic_to_a_topic_manager(self):
```
test_adding_new_topic_manager_to_a_topic

DubeySandeep replied:
Done!

seanlip commented on core/controllers/blog_homepage.py:
```diff
@@ -143,7 +143,8 @@ def get(self, author_username):
                 Exception(
                     'User with given username does not exist'))
-        if user_settings.role not in [BLOG_ADMIN, BLOG_POST_EDITOR]:
+        if not any([role in user_settings.roles for role in [
```
Can drop the '[' and ']' around "role ... BLOG_POST_EDITOR"

DubeySandeep replied:
Done!

seanlip commented on core/templates/pages/admin-page/roles-tab/admin-roles-tab.component.spec.ts:
```diff
@@ -136,36 +133,72 @@ describe('Admin roles tab component ', function() {
+      expect(component.startedEditingRole).toBeTrue();
+    });
+
+    it('should fetch user roles and intialize properties', fakeAsync(() => {
```
Spelling: initialize

DubeySandeep replied:
Fixed!

seanlip commented on core/templates/pages/admin-page/roles-tab/admin-roles-tab.component.ts:
```diff
@@ -16,170 +16,169 @@
+  }
+
+  markUserBanned(): void {
+    this.changingBannedVlaue = true;
```
Value is misspelled

bannedStatusChangeInProgress perhaps

DubeySandeep replied:
Fixed!

seanlip commented on core/templates/pages/admin-page/roles-tab/admin-roles-tab.component.ts:
```diff
@@ -16,170 +16,169 @@
+  ) {}
+
+  startEditing(): void {
+    this.startedEditingRole = true;
```
roleIsCurrentlyBeingEdited

DubeySandeep replied:
Done!

seanlip commented on core/templates/pages/admin-page/roles-tab/admin-roles-tab.component.ts:
```diff
@@ -16,170 +16,169 @@
+  showNewRoleSelector(): void {
+    this.possibleRolesToAdd = Object.keys(this.UPDATABLE_ROLES).filter(
+      role => !this.userRoles.includes(role)).sort();
+    this.selectingNewRole = true;
```
This seems a bit odd since no selecting is actually happening at the moment. Perhaps roleSelectorIsShown or similar?

DubeySandeep replied:
Renamed!

seanlip commented on core/templates/pages/admin-page/roles-tab/admin-roles-tab.component.ts:
```diff
@@ -16,170 +16,169 @@
+  }
+
+  removeRole(roleToRemove: string): void {
+    this.roleInUpdate = roleToRemove;
```
Why do we need to store this on the component? (I also don't really know what roleInUpdate means.)

DubeySandeep replied:
As we are going to show spinner around the role which we are adding/removing (i.e, in update).
roleInUpdate: The name of the role which is getting updated to user role i.e, either push or pop from users.roles field. Is it fine, I wasn't able to find a better name for this, do you have any suggestion?

seanlip replied:
OK, maybe try this.roleCurrentlyBeingUpdatedInBackend. 

Also maybe add a comment wherever this is initialized explaining that this is for both additions and removals. Do we null it out once the update has completed?

DubeySandeep replied:
Done!
> Do we null it out once the update has completed?

Yes, we are doing it on line 90!

seanlip commented on core/templates/pages/admin-page/roles-tab/topic-manager-role-editor-modal.component.ts:
```diff
@@ -36,34 +38,47 @@ export class TopicManagerRoleEditorModalComponent implements OnInit {
       this.updateTopicIdsForSelection();
+    }, data => {
+      let topicIdIndex = this.managerInTopicsWithId.indexOf(this.newTopicId);
+      this.managerInTopicsWithId.splice(topicIdIndex, 1);
```
managedTopicIds perhaps? I find managerInTopicsWithId hard to parse.

DubeySandeep replied:
Done!

seanlip commented on core/templates/pages/admin-page/roles-tab/admin-roles-tab.component.ts:
```diff
@@ -16,170 +16,169 @@
+      topicSummary => topicIdToName[topicSummary.id] = topicSummary.name);
+    modalRef.componentInstance.topicIdToName = topicIdToName;
+    modalRef.result.then(managerInTopicsWithId => {
+      this.managerInTopicsWithId = managerInTopicsWithId;
```
managedTopicIds perhaps?

DubeySandeep replied:
Done!

seanlip commented on core/tests/protractor_desktop/collections.js:
```diff
@@ -40,22 +38,17 @@ describe('Collections', function() {
-    await adminPage.addRole(PLAYER_USERNAME, 'collection editor');
-    await users.logout();
+    await users.createCollectionEditor(
+      'player@collections.com', PLAYER_USERNAME);
```
Why are we adding a "player" as a collection editor? I'm not sure that makes sense...

DubeySandeep replied:
We have this as of now (see line 57 in old file), I tried changing the email but I realised we have other collection editor below as well.

seanlip commented on core/templates/pages/admin-page/roles-tab/admin-roles-tab.component.html:
```diff
@@ -92,14 +92,14 @@ <h2>Role editor</h2>
         </button>
       </div>
-      <div *ngIf="!username">Please fillout the username above to editor the user role.</div>
+      <div *ngIf="!username">Please fillout the username above to edit the user role.</div>
```
Still not fixed. Start with "Please fill in..."

DubeySandeep replied:
From last comment I thought you were asking just to change editor to edit! Changed `fillout` to `fill in`

seanlip reviewed (changes requested):
Hi @DubeySandeep -- took a pass and left comments.

For the general comment, the last suggestion you made seems reasonable. Thanks!

DubeySandeep committed:
Addres review comments

DubeySandeep commented:
https://github.com/oppia/oppia/pull/12989#discussion_r673725691

@seanlip I've created an issue #13447 and I'll fix it in a seperate PR as it's not related to this PR i.e, we had a structure to add `is_topic_mange`, `is_moderator` etc. in the response, later we introduce userInfoHandler which provides infor about user role but we never remove the old structure. (Also, added a TODO!)

DubeySandeep committed:
Fix e2e test issue

seanlip commented:
Just two more comments:
- https://github.com/oppia/oppia/pull/12989#discussion_r674526833
- https://github.com/oppia/oppia/pull/12989#discussion_r674528874

DubeySandeep committed:
FIx test issue

kevintab95 reviewed (approved):
LGTM for codeowner files. Thanks @DubeySandeep!

DubeySandeep committed:
Address review comments

sagangwee reviewed (approved):
@DubeySandeep LGTM for codeowners, thanks!

seanlip commented on core/templates/pages/admin-page/roles-tab/admin-roles-tab.component.ts:
```diff
@@ -71,7 +76,9 @@ export class AdminRolesTabComponent implements OnInit {
 
   removeRole(roleToRemove: string): void {
-    this.roleInUpdate = roleToRemove;
+    // Updating roleCurrentlyBeingUpdatedInBackend as the given roleToRemove
```
I didn't mean to add the comment here. Add it above line 158 to explain what the property represents, then you can delete this one and the one on line 120.

DubeySandeep replied:
Done!

rohitkatlaa reviewed (approved):
LGTM for codeowner files. Thanks!

DubeySandeep committed:
Add code comment in correct place

DubeySandeep committed:
Remove unwanted cod comment

seanlip commented on core/templates/pages/admin-page/roles-tab/admin-roles-tab.component.ts:
```diff
@@ -40,6 +40,9 @@ export class AdminRolesTabComponent implements OnInit {
   managedTopicIds = [];
+  // The roleCurrentlyBeingUpdatedInBackend holds the role which is either being
+  // removed or added to user roles. This value is used to present a progress
+  // spinner around the role which is currently being updated in the backend.
```
Should "around" be "next to"?

DubeySandeep replied:
Done!

seanlip reviewed (approved):
Looks good, just one nit. Thanks!

DubeySandeep committed:
Nit fix in code comment

nithusha21 reviewed (approved):
I think this looks amazing! Thanks Sandeep :)

PS: This also does fix an issue I faced last release where someone with admin role couldn't run jobs. I think now I'll be able to assign both admin and release coordinator role to the same person if needed.

aks681 commented on core/domain/topic_services.py:
```diff
@@ -1056,6 +1056,32 @@ def deassign_user_from_all_topics(committer, user_id):
+        topic_id: str. The ID of the topic.
+
+    Raises:
+        Exception. The committer does not have rights to modify a role.
```
Incorrect exception docstring?

aks681 reviewed (approved):
Lgtm as codeowner! One minor comment.

DubeySandeep committed:
Merge branch 'develop' of github.com:oppia/oppia into use-roles

vojtechjelinek commented on core/templates/domain/user/user-info.model.ts:
```diff
@@ -63,30 +63,30 @@ export class UserInfo {
   static createDefault(): UserInfo {
     return new UserInfo(
-      'GUEST', false, false, false, false, false, null, null, null, false);
+      ['GUEST'], false, false, false, false, false, null, null, null, false);
```
Do we have a constant for this role? Ditto for the roles below.

DubeySandeep replied:
Not in the frontend as of now, but we are planning to move all the roles constants to fronted which will simplify the UserInfo constructor as well. (Going to be @ashutoshc8101 next PR!)

vojtechjelinek replied:
Ok, thanks!

vojtechjelinek reviewed (changes requested):
Thanks! One comment.

vojtechjelinek reviewed (approved):
LGTM!

DubeySandeep committed:
Add test for modal component

DubeySandeep committed:
Fix test issue

DubeySandeep committed:
Adds modal test file

DubeySandeep committed:
Merge branch 'develop' of github.com:oppia/oppia into use-roles

DubeySandeep committed:
Fix test

DubeySandeep committed:
Fix test

DubeySandeep committed:
Resolve conflicts

DubeySandeep committed:
Fix test

DubeySandeep committed:
Fix test issue
